Request attachment icon again if needed

squelart opened the pull request:
#### 8852bc5adfdbc818376a780dc93ffac266e2e268
<pre>
Request attachment icon again if needed
<a href="https://bugs.webkit.org/show_bug.cgi?id=274951">https://bugs.webkit.org/show_bug.cgi?id=274951</a>
<a href="https://rdar.apple.com/128175175">rdar://128175175</a>

Reviewed by Aditya Keerthi.

The attachment icon was only requested once, during the first
layout&amp;painting.
But if the title and/or type attributes are changed, the icon
should be requested again to match these attributes.

In particular, this fixes an issue when sharing a document from
Keynote or other iWork app, where the attachment is initially
created with a generic &quot;application/octet-stream&quot; type, and later
updated to the proper type like &quot;com.apple.iwork.keynote.sffkey&quot;.

* LayoutTests/fast/attachment/attachment-icon-update-after-type-change-expected.html: Added.
* LayoutTests/fast/attachment/attachment-icon-update-after-type-change.html: Added.
* LayoutTests/fast/attachment/resources/attachment-test-utils.js:
(takeScreenshot):
(takeScreenshotWhenAttachmentsSettled):
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::setFile):
(WebCore::HTMLAttachmentElement::attributeChanged):
(WebCore::HTMLAttachmentElement::updateAttributes):
(WebCore::HTMLAttachmentElement::setNeedsIconRequest):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
(WebCore::HTMLAttachmentElement::requestIconIfNeededWithSize):
(WebCore::HTMLAttachmentElement::setNeedsWideLayoutIconRequest): Deleted.
(WebCore::HTMLAttachmentElement::requestIconWithSize): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::paintAttachmentIcon):

Canonical link: <a href="https://commits.webkit.org/279620@main">https://commits.webkit.org/279620@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8ca58f1bf0e57195636dac30dbcc8544327accb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53949 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33319 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57227 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4673 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56251 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40827 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4564 "Failed to checkout and rebase branch from PR 29371") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43680 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3082 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56046 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31531 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24821 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28364 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3992 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2825 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58820 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29124 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/4564 "Failed to checkout and rebase branch from PR 29371") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51096 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30316 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46811 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50434 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11764 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31267 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30091 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.h:
```diff
@@ -152,7 +152,7 @@ class HTMLAttachmentElement final : public HTMLElement {
     mutable RefPtr<DOMRectReadOnly> m_saveButtonClientRect;
 
-    bool m_needsWideLayoutIconRequest { false };
+    bool m_needsIconRequest { true };
```
Can you elaborate on why the default value is being changed?

squelart replied:
It used to only be set to true from `ensureWideLayoutShadowTree()` (so only for wide-layout attachments), but now that we want to also set it for non-wide layout -- hence the renaming, we may as well just set it to true during construction.

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -335,7 +335,6 @@ void HTMLAttachmentElement::ensureWideLayoutShadowTree(ShadowRoot& root)
 
     m_imageElement = createContainedElement<HTMLImageElement>(previewArea, attachmentIconIdentifier());
     AttachmentImageEventsListener::addToImageForAttachment(*m_imageElement, *this);
-    setNeedsWideLayoutIconRequest();
```
Is this why the default value of the member variable is being changed?

squelart replied:
It's more that the default value has changed (because we always want it true for all attachments), so we don't need this line here anymore.

pxlcoder reviewed (approved):
Ideally the bug/PR title would briefly describe the problem (user-facing issue) that is being solved.

squelart committed:
Request attachment icon again if needed
https://bugs.webkit.org/show_bug.cgi?id=274951
rdar://128175175

Reviewed by Aditya Keerthi.

The attachment icon was only requested once, during the first
layout&painting.
But if the title and/or type attributes are changed, the icon
should be requested again to match these attributes.

In particular, this fixes an issue when sharing a document from
Keynote or other iWork app, where the attachment is initially
created with a generic "application/octet-stream" type, and later
updated to the proper type like "com.apple.iwork.keynote.sffkey".

* LayoutTests/fast/attachment/attachment-icon-update-after-type-change-expected.html: Added.
* LayoutTests/fast/attachment/attachment-icon-update-after-type-change.html: Added.
* LayoutTests/fast/attachment/resources/attachment-test-utils.js:
(takeScreenshot):
(takeScreenshotWhenAttachmentsSettled):
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureWideLayoutShadowTree):
(WebCore::HTMLAttachmentElement::setFile):
(WebCore::HTMLAttachmentElement::attributeChanged):
(WebCore::HTMLAttachmentElement::updateAttributes):
(WebCore::HTMLAttachmentElement::setNeedsIconRequest):
(WebCore::HTMLAttachmentElement::requestWideLayoutIconIfNeeded):
(WebCore::HTMLAttachmentElement::requestIconIfNeededWithSize):
(WebCore::HTMLAttachmentElement::setNeedsWideLayoutIconRequest): Deleted.
(WebCore::HTMLAttachmentElement::requestIconWithSize): Deleted.
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/rendering/AttachmentLayout.mm:
(WebCore::AttachmentLayout::AttachmentLayout):
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::paintAttachmentIcon):

Canonical link: https://commits.webkit.org/279620@main
